src/core/thread/network_data_service.cpp

```diff
@@ -59,28 +59,15 @@ Error Manager::AddService(const void *aServiceData,
                           const void *aServerData,
                           uint8_t     aServerDataLength)
 {
-    Error error;
-
-    SuccessOrExit(error = Get<Local>().AddService(
-                      kThreadEnterpriseNumber, reinterpret_cast<const uint8_t *>(aServiceData), aServiceDataLength,
-                      aServerStable, reinterpret_cast<const uint8_t *>(aServerData), aServerDataLength));
-
-    Get<Notifier>().HandleServerDataUpdated();
-
-exit:
-    return error;
+    return Get<Local>().AddService(kThreadEnterpriseNumber, reinterpret_cast<const uint8_t *>(aServiceData),
```

Comment:
I think this code change is to avoid multiple sending multiple `NETDATA.ntf` for each small change. 

But this adds extra complexity to callers of `AddService` because now callers have to call `HandleServerDataUpdated` subsequently. 

Can we keep this method unchaged, but change `HandleServerDataUpdated` to **NOT** call `SynchronizeServerData` directly, but post a tasklet to call `SynchronizeServerData` ? 
This can merge multiple `AddService` or `RemoveService` to one `NETDATA.ntf`, and also simplify the logic of callers of `AddService`.

Thoughts?

Reply:
> Can we keep this method unchaged, but change HandleServerDataUpdated to NOT call SynchronizeServerData directly, but post a tasklet to call SynchronizeServerData ?

This is a good idea. I think we considered something similar this a long time ago but from I somewhat recall there were some corner cases that needed some care/thought to handle.

I think this is good thing to consider for a future enhancement/PR (it would be a bigger behavior change of `NetworkData::Notifier`).

----

In the first version of this change I had a paramter `aRegisterWithLeader` in `AddService()`, ... which indicated whether the `AddService()` itself should do the registeration with leader or leave it for the caller to do. This made it somewhat more complex to use. I changed and simplified the `AddService()` per @jwhui  suggestion from https://github.com/openthread/openthread/pull/6587#discussion_r633219552 which make it consistent with other  network data entries such as on-mesh prefixes and routes.

Reply:
Posting a tasklet works well when the APIs are being called from the same tasklet. However, posting a tasklet does not handle the case where APIs are being invoked as a result of CLI (or spinel in the case of NCP) interaction, even when the CLI is being driven by a script.

Reply:
> Posting a tasklet works well when the APIs are being called from the same tasklet. However, posting a tasklet does not handle the case where APIs are being invoked as a result of CLI (or spinel in the case of NCP) interaction, even when the CLI is being driven by a script.

I don't understand why. Could you give more explanation?

Calling `SynchronizeServerData` in a tasklet allows multiple changes using `AddService` or `RemoveService` to be merged in one `NETDATA.ntf`, right after this loop.

Reply:
The tasklet model only consolidates multiple updates when `AddService` and `RemoveService` are called from the same task context. When issuing commands via CLI or spinel that invoke calls to `AddService` and `RemoveService` via CLI or spinel, there is no way to guarantee that such calls will happen in the same task context.

Reply:
> The tasklet model only consolidates multiple updates when AddService and RemoveService are called from the same task context.

Yes. But isn't this solve our problem? Currently we are calling multiple AddService/RemoveService in the same call stack (and hense same Tasklet) and use `registerWithLeader` to track if we need to call `HandleServerDataUpdated`. But if we just post a tasklet for running `HandleServerDataUpdated` whenever we call AddService/RemoveService, we don't need to use an additional boolean (i.e. registerWithLeader) to track if we need to call `HandleServerDataUpdated`, which will simplify our code. It also makes calling AddService/RemoveService simpler, not having to call  `HandleServerDataUpdated` afterwards.

Reply:
I don't think we necessarily achieve the same benefit if someone is invoking `AddPrefix` and `RemovePrefix` using the CLI. How do we guarantee that the `prefix add` and `prefix remove` commands will be sent and processed within the same task context?